Sources/TuistCache/Cache/CacheRemoteStorage.swift

```diff
@@ -0,0 +1,114 @@
+import Foundation
+import RxSwift
+import TSCBasic
+import TuistCore
+import TuistSupport
+
+enum CacheRemoteStorageError: LocalizedError {
+    case missingCloudConfig
+    case incorrectCloudConfig
+}
+
+struct CloudHEADResponse: Decodable {}
+
+// TODO: Later, add a warmup function to check if it's correctly authenticated ONCE
+final class CacheRemoteStorage: CacheStoring {
+    // MARK: - Attributes
+
+    private let cloudClient: CloudClienting
+
+    // MARK: - Init
+
+    init(cloudClient: CloudClienting) {
+        self.cloudClient = cloudClient
+    }
+
+    // MARK: - CacheStoring
+
+    func exists(hash: String, userConfig: Config) -> Single<Bool> {
+        let resource = existsResource(hash: hash, userConfig: userConfig)
+        return cloudClient.request(resource).map { response in
+            let successRange = 200 ..< 300
+            return successRange.contains(response.response.statusCode)
+        }
+    }
+
+    func fetch(hash: String, userConfig: Config) -> Single<AbsolutePath> {
+        let resource = fetchResource(hash: hash, userConfig: userConfig)
+        return cloudClient.request(resource).map { _ in
+            AbsolutePath.root // TODO:
+        }
+    }
+
+    func store(hash: String, userConfig: Config, xcframeworkPath _: AbsolutePath) -> Completable {
+        let resource = storeResource(hash: hash, userConfig: userConfig)
+        return cloudClient.request(resource).map { responseTuple in
+            let cacheResponse = responseTuple.object.data
+            let artefactToDownloadURL = cacheResponse.url
+            print(artefactToDownloadURL)
+            // TODO: Download file at given url
+        }.asCompletable()
+    }
+
+    // MARK: - Fileprivate
+
+    fileprivate func apiCacheURL(hash: String, userConfig: Config, contentMD5: String? = nil) throws -> URL {
+        guard let cloudConfig = userConfig.cloud else { throw CacheRemoteStorageError.missingCloudConfig }
+        guard var urlComponents = URLComponents(url: cloudConfig.url, resolvingAgainstBaseURL: false) else {
+            throw CacheRemoteStorageError.incorrectCloudConfig
+        }
+
+        urlComponents.path = "api/cache"
+        var queryItems = [
+            URLQueryItem(name: "project_id", value: cloudConfig.projectId),
+            URLQueryItem(name: "hash", value: hash),
+        ]
+
+        if let contentMD5 = contentMD5 {
+            queryItems.append(URLQueryItem(name: "content_md5", value: contentMD5))
+        }
+
+        urlComponents.queryItems = queryItems
+        return urlComponents.url!
+    }
+
+    fileprivate func existsRequest(hash: String, userConfig: Config) throws -> URLRequest {
+        var urlRequest = URLRequest(url: try apiCacheURL(hash: hash, userConfig: userConfig))
+        urlRequest.httpMethod = "HEAD"
+        return urlRequest
+    }
+
+    fileprivate func fetchRequest(hash: String, userConfig: Config) throws -> URLRequest {
```

Comment:
Considering the `__Request` methods are only used from one place and that they are very short, wouldn't it make sense to merge them into the `___Resource` methods?

Reply:
In fact, what about moving them fro this class. For example, to check whether an artifact exists or not in the cache, we could do:

```swift
client.execute(CloudCacheResponse.exists)
```

Reply:
ok I moved these functions to their respective response classes